From 8ebac804e964a23aa6b1342653b16baf3143704a Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:17:22 -0400
Subject: [net] netfilter: nf_tables: Stack expression type depending on their family

Message-id: <20141021161722.22888.32683.stgit@localhost.localdomain>
Patchwork-id: 97936
O-Subject: [RHEL7 net PATCH 038/124] netfilter: nf_tables: Stack expression type depending on their family
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 758dbcecf180a161e15971a2674fa6e65622a281

 commit 758dbcecf180a161e15971a2674fa6e65622a281
 Author: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>
 Date:   Mon Apr 14 15:41:26 2014 +0300

    netfilter: nf_tables: Stack expression type depending on their family

    To ensure family tight expression gets selected in priority to family
    agnostic ones.

    Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index 0911131..3835346 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -1093,7 +1093,10 @@ static void nft_ctx_init(struct nft_ctx *ctx,
 int nft_register_expr(struct nft_expr_type *type)
 {
  nfnl_lock(NFNL_SUBSYS_NFTABLES);
- list_add_tail(&type->list, &nf_tables_expressions);
+ if (type->family == NFPROTO_UNSPEC)
+  list_add_tail(&type->list, &nf_tables_expressions);
+ else
+  list_add(&type->list, &nf_tables_expressions);
  nfnl_unlock(NFNL_SUBSYS_NFTABLES);
  return 0;
 }
-- 
1.7.1